127 128 131 133 fast api selenium celery

fracotti opened the pull request:
This pull request dedicated to tasks:
#127 - FastAPI
#128 - FastAPI EDRPOU/RNOKPP validation endpoint subtask
#131 - Selenium
#133 - Celery

Additionally was implemented Redis and Redis Commander

fracotti committed:
Moved main.py to app folder. Deleted logger from main.py. FastAPi setup configured.

fracotti committed:
Moved settings.py to app folder.

fracotti committed:
Created tasks for Celery and Selenium.

fracotti committed:
Created route for FastAPI to validate EDRPOU/RNOKPP codes. __init__.py for /app

fracotti committed:
Created and configured celery

fracotti committed:
Added and configured Dockerfile, docker-compose.utils.yml and docker-compose.utils_dev.yml (for developing tools such as redis-commander)

fracotti committed:
Added short instruction how to configure .env file and how to build and run microservice

fracotti committed:
Added requirements of needed packages for Utils & Validation microservice

fracotti committed:
Added code for validation via selenium.

fracotti committed:
Merge remote-tracking branch 'origin/127-128-131-133-FastAPI-Selenium-Celery' into fastapi

VovaD199 commented on BackEnd/utils_validation_service/app/selenium_validation.py:
```diff
@@ -0,0 +1,50 @@
+        return f'Error {e2}'
+    finally:
+        driver.quit()
+    return status
```
**Finally** calls driver.quit(), which is good.

But if WebDriverWait fails with an error, the code will exit before quit(), leaving Chrome processes "hanging".
Solution: Use try inside finally:

finally:
    try:
        driver.quit()
    except Exception as e:
        logger.error(f"Error while closing driver: {e}")

fracotti replied:
It makes sense. I will change that

VovaD199 commented on BackEnd/utils_validation_service/app/selenium_validation.py:
```diff
@@ -0,0 +1,50 @@
+            logger.info(f'{code} is {status.lower()}')
+        except Exception as e1:
+            logger.error(e1)
+            pass
```
Dangerous result check:
try:
    result_element = WebDriverWait(driver, 5).until(
        ec.presence_of_element_located((By.CSS_SELECTOR, "h4"))
    )
    if "контрагентів не знайдено" in result_element.text:
        status = 'Invalid'
except Exception:
    pass

#Now even if h4 disappears or changes, the code will not crash.

VovaD199 commented on BackEnd/utils_validation_service/app/selenium_validation.py:
```diff
@@ -0,0 +1,50 @@
+def validate_code(code: str):
+    driver = get_selenium_driver()
+    try:
+        driver.get("https://youcontrol.com.ua/")
```
If the site changes address, you will need to search and change the URL throughout the code.
YOUCONTROL_URL = "https://youcontrol.com.ua/"

def validate_code(code: str):
    driver = get_selenium_driver()
    try:
        driver.get(YOUCONTROL_URL)

fracotti replied:
It is a nice idea

Maksym04u reviewed (approved):
Good job.

fracotti committed:
Reconfigured logger to log info about code validation and status of validation.
Added info_file handler to settings.py
Set log level to INFO

fracotti committed:
Implemented changes and recommendations from Volodymyr.
Added if-block to check if specific text that mean that there is no data about our request in YouControl database is in the h4 tag
Store YouControl url in constant.
